[WIP][master] Greatly Simplify(refactor) Salt's Logging

s0undt3ch opened the pull request:
### What does this PR do?
This PR simplifies Salt's logging setup, in particular it changes how we handle multiprocessing logging.
It also deprecates the old logging setup and set's it's removal for the first release after 2022/Jan/01.

### Previous Behavior
Multiprocessing worked OK in Linux, and not so OK in Windows.

### New Behavior
Logging setup simplified and multiprocessing logging working great on all distributions

### Tests written?

Yes, and some fixed.

### Commits signed with GPG?

Yes

dwoz committed:
Remove existing handlers in sub-processes

dwoz reviewed (changes requested):
I'd like to see this PR broken up into several smaller and more digestible changes.

waynew reviewed (changes requested):
Looks, great, just a couple of comments and a typo

waynew commented on salt/_logging/__init__.py:
```diff
@@ -0,0 +1,18 @@
+    As the name suggests, this is considered an internal API which can change without notice,
+    although, as best effort, we'll try not to break code relyin on it.
+
+    The ``salt._logging`` package should be imported as soon as possible since salt teaks
```
oh, tweaks?

s0undt3ch replied:
Yes, tweaks :)

waynew commented on tests/support/helpers.py:
```diff
@@ -1656,3 +1657,35 @@ def dedent(text, linesep=os.linesep):
     return clean_text
+
+
+class CaptureSTDs(object):
```
There *may* be a better name than `CaptureSTDs`. `CaptureOutput`? Maybe two decorators, a `capture_stdout` and a `capture_stderr`?

s0undt3ch replied:
The same class could be made to optionally capture each of the outputs.

:+1: on the name change

s0undt3ch replied:
FYI, pytest has this out of the box :)

s0undt3ch replied:
Addressed

s0undt3ch commented on salt/_logging/impl.py:
```diff
@@ -0,0 +1,1438 @@
+        try:
+            max_logger_length = len(max(
+                list(logging.Logger.manager.loggerDict), key=len
+            ))
```
That was carried over from the previous implementation, I'm sure there was a reason, but I'll investigate.

s0undt3ch replied:
Consider the following code:
```python
        import timeit
        t1_max_logger_length = len(max(
            list(logging.Logger.manager.loggerDict), key=len
        ))
        t1 = timeit.timeit('ParserBase.t1()', setup='from tests.unit.utils.test_parsers import ParserBase')
        t2_max_logger_length = max(
            len(name) for name in logging.Logger.manager.loggerDict
        )
        t2 = timeit.timeit('ParserBase.t2()', setup='from tests.unit.utils.test_parsers import ParserBase')
        print('t1', t1_max_logger_length, t1)
        print('t2', t2_max_logger_length, t2)
        t1_max_logger_length = len(max(
            list(logging.Logger.manager.loggerDict), key=len
        ))
        t1 = timeit.timeit('ParserBase.t1()', setup='from tests.unit.utils.test_parsers import ParserBase')
        t2_max_logger_length = max(
            len(name) for name in logging.Logger.manager.loggerDict
        )
        t2 = timeit.timeit('ParserBase.t2()', setup='from tests.unit.utils.test_parsers import ParserBase')
        print('t1', t1_max_logger_length, t1)
        print('t2', t2_max_logger_length, t2)


    @staticmethod
    def t1():
        max_logger_length = len(max(
            list(logging.Logger.manager.loggerDict), key=len
        ))

    @staticmethod
    def t2():
        max_logger_length = max(
            len(name) for name in logging.Logger.manager.loggerDict
        )
```
Now, the outcome of that:
```
t1 27 8.73357391357                                                                                             
t2 29 12.3908040524                                                                                                                                                                                                              
t1 29 8.77210521698                                                                                                                                                                                                              
t2 29 12.4122331142
```
On the first iteration, the max length wasn't equal, maybe another logger was added, so let's ignore that on the first run, specially since on the second `timeit` they both report the same length.
Now, the existing implementation is tad faster than your proposed solution.
Maybe I reached this same conclusion the first time around.

s0undt3ch committed:
Close handlers

s0undt3ch committed:
Reset some more attributes

s0undt3ch commented:
The macos failures(pipeline timeout, not SSH timeout) look valid and require more inspection.

s0undt3ch commented on salt/utils/process.py:
```diff
@@ -402,31 +402,42 @@ def add_process(self, tgt, args=None, kwargs=None, name=None):
-                        kwargs['log_queue_level'] = (
-                            salt.log.setup.get_multiprocessing_logging_level()
-                        )
+        # Need to ensure that 'log_queue' and 'log_queue_level' is
```
Rework comment.

s0undt3ch commented on salt/utils/process.py:
```diff
@@ -402,31 +402,42 @@ def add_process(self, tgt, args=None, kwargs=None, name=None):
+        if type(Process) is type(tgt) and issubclass(tgt, Process):
+            if 'log_host' not in kwargs:
+                try:
+                    salt._logging.set_log_forwarding_host(self.log_host)
```
Set it after forking?

s0undt3ch commented on salt/utils/process.py:
```diff
@@ -684,36 +695,53 @@ def kill_children(self, *args, **kwargs):
+        if log_host is None:
+            log_host = salt._logging.get_log_forwarding_host()
+        else:
+            # Set the logging queue so that it can be retrieved later
```
Fix comment

s0undt3ch committed:
Set the root logger level to the right one once we have opts available.

dwoz committed:
Add ZMQ to process name

dwoz committed:
Fix filename test

dwoz committed:
Add more time for connection

dwoz committed:
Fix up test case

dwoz committed:
Fix up things based on PR review comments

s0undt3ch committed:
Fix module reference under windows

s0undt3ch committed:
Move `__global_logging_exception_handler` to `salt._logging`

s0undt3ch committed:
Prevent the log forwarding handler from being inherited by sub-processes

s0undt3ch committed:
Skip on `AssrtionError` when running under CI, for now.

These tests pass locally.

s0undt3ch committed:
Deprecate the old logging setup until 2023-01-01

s0undt3ch committed:
Treat `SaltKeyOptionParser` as a first class logging citizen

s0undt3ch committed:
Make sure the passed in `*args` and `kwargs` are always captured

s0undt3ch committed:
Allow registering after fork and finalize methods

s0undt3ch commented:
re-run full all

s0undt3ch committed:
Use separate ZMQ Context's

s0undt3ch commented:
re-run full all

s0undt3ch committed:
Override `Handler.close` to make sure it also stops the handler

s0undt3ch committed:
Allow the deferred logging tests to run on windows

s0undt3ch commented:
re-run full all

s0undt3ch committed:
Wait a little longer when stopping the handler

s0undt3ch commented:
re-run full all

s0undt3ch committed:
Limit logger names on formatted logs to 80 chars by default.

When calculating the maximum logger name length, make sure it's not longer than 80 chars.
Messages which need more that 80 chars will use them and be properly formatted, but not ALL log messages.

s0undt3ch committed:
Initial commit of zmq based multiprocessing logging

s0undt3ch committed:
Do not leave socket open on windows

s0undt3ch committed:
Add the new refactored salt logging.

It also handles multiprocessing logging.

s0undt3ch committed:
Tweak logging as soon as salt is imported

s0undt3ch committed:
Update the parsers utility code to the new logging setup

s0undt3ch committed:
Adjust Salt's CLI scripts to the new parser and logging changes

s0undt3ch committed:
The `Process` class will now take care of all `__getstate__` and `__setstate__`

s0undt3ch committed:
Adapt salt code the the new logging

s0undt3ch committed:
Tests blow up on windows with our `mock_open` implementation.

s0undt3ch committed:
Stop using ``NewStyleClassMixin`` on logging

s0undt3ch commented:
The proposed changes are now merged into master.
